qcp/gates.py

```diff
@@ -0,0 +1,149 @@
+import cmath
+
+from matrix import Matrix
+import constants as c
+from tensor_product import tensor_product
+
+
+# Notation note : |001> represents a 3 qubit system where the first qubit is |1> and the second and third qubit is |0>
+# Targets and Controls work off this notation
+
+def multi_gate(size: int, targets: list[int], gate: str, phi=complex(0)):
+    """
+    Constructs a (2**size by 2**size) gate matrix that applies a specific gate to one or more specified qubits
+
+    :param size: total number of qubits in circuit -> int
+    :param targets: list of qubits the specified gate will be applied to -> List[int]
+    :param gate: string character representing which specified gate we want to apply;
+                                h = hadamard, x = Pauli x, z = Pauli z, p = Phase gate
+    :param phi: Phase angle for the phase gate -> complex number
+    :return: Matrix([int])
+    """
+
+    if gate == "h":
+        g = c.TWO_HADAMARD
+    elif gate == "x":
+        g = c.PAULI_X
+    elif gate == "z":
+        g = c.PAULI_Z
+    elif gate == "p":
+        g = phase_shift(phi)
+    else:
+        return
+
+    m = Matrix([1])
+    t = [x - 1 for x in targets]
```

Comment:
Is `targets` in this case a list of qubit numbers to target, so say for example `[0,1,4]` against a qubit `|01010>`?